File: v29_state/java/Game2.java
package v29_state.java;

interface State {
    void walk();
    void run();
    void jump();
    void idle();
}

class IdleState implements State {
    private Character character;

    public IdleState(Character character) {
        this.character = character;
    }

    public void walk() {
        character.setState(character.getWalkingState());
        System.out.println("The character is now walking.");
    }

    public void run() {
        System.out.println("The character cannot run right now.");
    }

    public void jump() {
        System.out.println("The character cannot jump right now.");
    }

    public void idle() {
        System.out.println("The character is already idle.");
    }
}

class WalkingState implements State {
    private Character character;

    public WalkingState(Character character) {
        this.character = character;
    }

    public void walk() {
        System.out.println("The character is already walking.");
    }

    public void run() {
        character.setState(character.getRunningState());
        System.out.println("The character is now running.");
    }

    public void jump() {
        System.out.println("The character cannot jump right now.");
    }

    public void idle() {
        character.setState(character.getIdleState());
        System.out.println("The character is now idle.");
    }
}

class RunningState implements State {
    private Character character;

    public RunningState(Character character) {
        this.character = character;
    }

    public void walk() {
        System.out.println("The character cannot walk right now.");
    }

    public void run() {
        System.out.println("The character is already running.");
    }

    public void jump() {
        character.setState(character.getJumpingState());
        System.out.println("The character is now jumping.");
    }

    public void idle() {
        character.setState(character.getIdleState());
        System.out.println("The character is now idle.");
    }
}

class JumpingState implements State {
    private Character character;

    public JumpingState(Character character) {
        this.character = character;
    }

    public void walk() {
        System.out.println("The character cannot walk right now.");
    }

    public void run() {
        System.out.println("The character cannot run right now.");
    }

    public void jump() {
        System.out.println("The character is already jumping.");
    }

    public void idle() {
        character.setState(character.getIdleState());
        System.out.println("The character is now idle.");
    }
}

class Character {
    private State idleState;
    private State walkingState;
    private State runningState;
    private State jumpingState;

    private State state;

    public Character() {
        idleState = new IdleState(this);
        walkingState = new WalkingState(this);
        runningState = new RunningState(this);
        jumpingState = new JumpingState(this);

        state = idleState;
    }

    public void walk() {
        state.walk();
    }

    public void run() {
        state.run();
    }

    public void jump() {
        state.jump();
    }

    public void idle() {
        state.idle();
    }

    public void setState(State state) {
        this.state = state;
    }

    // Getter methods for the states
    public State getIdleState() {
        return idleState;
    }

    public State getWalkingState() {
        return walkingState;
    }

    public State getRunningState() {
        return runningState;
    }

    public State getJumpingState() {
        return jumpingState;
    }
}


public class Game2 {
    public static void main(String[] args) {
        Character character = new Character();

        character.walk(); // The character is now walking.
        character.run(); // The character is now running.
        character.jump(); // The character is now jumping.
        character.idle(); // The character is now idle.

        character.run(); // The character cannot run right now.
        character.jump(); // The character cannot jump right now.
    }
}

File: v29_state/python/game2.py
class State:
    def walk(self):
        pass

    def run(self):
        pass

    def jump(self):
        pass

    def idle(self):
        pass


class IdleState(State):
    def __init__(self, character):
        self.character = character

    def walk(self):
        self.character.state = self.character.walking_state
        print("The character is now walking.")

    def run(self):
        print("The character cannot run right now.")

    def jump(self):
        print("The character cannot jump right now.")

    def idle(self):
        print("The character is already idle.")


class WalkingState(State):
    def __init__(self, character):
        self.character = character

    def walk(self):
        print("The character is already walking.")

    def run(self):
        self.character.state = self.character.running_state
        print("The character is now running.")

    def jump(self):
        print("The character cannot jump right now.")

    def idle(self):
        self.character.state = self.character.idle_state
        print("The character is now idle.")


class RunningState(State):
    def __init__(self, character):
        self.character = character

    def walk(self):
        print("The character cannot walk right now.")

    def run(self):
        print("The character is already running.")

    def jump(self):
        self.character.state = self.character.jumping_state
        print("The character is now jumping.")

    def idle(self):
        self.character.state = self.character.idle_state
        print("The character is now idle.")


class JumpingState(State):
    def __init__(self, character):
        self.character = character

    def walk(self):
        print("The character cannot walk right now.")

    def run(self):
        print("The character cannot run right now.")

    def jump(self):
        print("The character is already jumping.")

    def idle(self):
        self.character.state = self.character.idle_state
        print("The character is now idle.")


class Character:
    def __init__(self):
        self.idle_state = IdleState(self)
        self.walking_state = WalkingState(self)
        self.running_state = RunningState(self)
        self.jumping_state = JumpingState(self)

        self.state = self.idle_state

    def walk(self):
        self.state.walk()

    def run(self):
        self.state.run()

    def jump(self):
        self.state.jump()

    def idle(self):
        self.state.idle()


if __name__ == "__main__":
    character = Character()

    character.walk()  # The character is now walking.
    character.run()  # The character is now running.
    character.jump()  # The character is now jumping.
    character.idle()  # The character is now idle.

    character.run()  # The character cannot run right now.
    character.jump()  # The character cannot jump right now.


